Permanent magnet branch merge

akaptano opened the pull request:
Hi all,

The permanent magnet branch is looking pretty ready for a merge into the main branch. Some known issues/idiosyncrasies:
1. I am not quite saving the PermanentMagnetGrid class correctly with the JSON/Optimizable stuff, so I hope to get some feedback on that.
2. I have made an examples/4_Paper_Reproductions/ folder to hold some of the full-scale, high-resolution permanent magnet run files and files for reproducing my papers. I know this is not typical, but I am wondering if there is a solution other than keeping these files outside the SIMSOPT repo/on another branch, since I worry that they will stagnate relative to the rest of the SIMSOPT code (which develops fast!)
3. The unit tests for the magnet functionality are relatively thin. This is something I'm happy to address in more detail when I get the chance. 

Looking forward to any other thoughts, and thanks for the help!

mbkumar commented:
Thank you Alan for the contribution!

> 
> 1. I am not quite saving the PermanentMagnetGrid class correctly with the JSON/Optimizable stuff, so I hope to get some feedback on that.
Could you please elaborate on this point? 

> 2. I have made an examples/4_Paper_Reproductions/ folder to hold some of the full-scale, high-resolution permanent magnet run files and files for reproducing my papers. I know this is not typical, but I am wondering if there is a solution other than keeping these files outside the SIMSOPT repo/on another branch, since I worry that they will stagnate relative to the rest of the SIMSOPT code (which develops fast!)
 I would set them up in a separate repo and develop scaled-down examples to be included in simsopt. You can refer the readers to the separate repo in the example files. The code in the examples is fairly stable despite the big changes to the source code. So no need to worry about stagnation. Don't forget to pin the simsopt version for your separate repo, so users could run those files always.

> 3. The unit tests for the magnet functionality are relatively thin. This is something I'm happy to address in more detail when I get the chance.
Yes. Many newly added functions do not have unit tests. Please do this at the earliest, otherwise, we do not want to merge the code.

> 
> Looking forward to any other thoughts, and thanks for the help!
I will go through the changes sometime this week and get back with more detailed comments.

Once again thanks for the contribution. This is a big one.

landreman commented:
@ejpaul it looks like this PR would bring in your bounce averaging work. What do you think about this?

akaptano commented:
@landreman This is correct, although I merged in the bounce branch from @ejpaul  a while back, so may need another update to be representative of the current state of the bounce branch.

akaptano commented:
@mbkumar Here are some responses:
(1) A while back, I added the as_dict and from_dict functions to the PermanentMagnetClass, but it looks like the syntax/exact functions for saving the JSON files has changed a bit since then. So I was previously able to save a PermanentMagnetClass but no longer. Part of the issue is that this class depends in turn on additional class objects with their own as_dict and from_dict functions.
(2) I will get working on moving over the big examples
(3) I will get working on the unit tests, although with Simons and some interviews coming up, this might be a few weeks before I can really address the lack here. 

Thanks for the help!

mbkumar commented:
@akaptano 
I can help you with the JSON. How about we meet sometime next week?

mbkumar commented:
Sure. Lets do that.

Bharat Medasani

Engineer
Princeton Plasma Physics Lab (PPPL)


On Wed, Mar 8, 2023 at 12:09 PM Alan Kaptanoglu ***@***.***>
wrote:

> @mbkumar <https://github.com/mbkumar> I will likely be very busy next
> week -- if you will be at the Simons meeting at Princeton, it might be
> easier to get together in person there.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hiddenSymmetries/simsopt/pull/296#issuecomment-1460516886>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA62VEE7KQK2RLCGHAR4PQ3W3C4LHANCNFSM6AAAAAAVTJER6I>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

akaptano commented:
@landreman @ejpaul I have removed the bounce branch functionality from this PR since it is not a necessary addition and I don't have unit tests for this anyways.

landreman commented on src/simsopt/field/magneticfieldclasses.py:
```diff
@@ -502,6 +503,181 @@ def from_dict(cls, d, serial_objs_dict, recon_objs):
         return field
 
 
+class DipoleField(MagneticField):
```
The magneticfieldclasses.py file is pretty big already, and should probably be split up. What do you think of putting `DipoleField` in its own file?

landreman commented on src/simsoptpp/python_surfaces.cpp:
```diff
@@ -149,6 +149,7 @@ template <typename T, typename S> void register_common_surface_methods(S &s) {
      .def("fit_to_curve", &T::fit_to_curve, py::arg("curve"), py::arg("radius"), py::arg("flip_theta") = false)
      .def("scale", &T::scale)
      .def("extend_via_normal", &T::extend_via_normal)
+     .def("extend_via_projected_normal", &T::extend_via_projected_normal)
```
Somewhere, perhaps here so it shows up as a python docstring, can you add a comment with a few sentences explaining what `extend_via_projected_normal` does, and what its arguments mean?

landreman replied:
Presently these comments are not visible from python. But if you move these comments to a string inside the closing parenthesis of the `.def()` bit in the line above, then they become regular python docstrings visible to IDEs and `help()`. See e.g. `python_boozermagneticfield.cpp`

landreman reviewed (commented):
Here's a first batch of comments - more to come when I have more time to study the code.

landreman commented on src/simsopt/util/adjust_magnet_angles.py:
```diff
@@ -0,0 +1,699 @@
+"""
+
+import numpy as np
+import pandas as pd
```
This line introduces a dependency on pandas. Every new dependency is an opportunity for installation to fail, so we should look carefully at this. Also, some of the unit tests are failing due to this import. It appears that pandas is only used in the function `moments_file()`, which is never called. Can we delete that function and so remove this import of pandas?

landreman commented on src/simsopt/util/adjust_magnet_angles.py:
```diff
@@ -0,0 +1,699 @@
+
+if __name__ == '__main__':
+
+    #dirname = '/u/khammond/focus/ncsx/magnets_trial24_single/'
+    #filename = dirname + 'trial_24_single.focus'
```
Remove commented-out code

landreman reviewed (commented):
Some unit tests are failing due to not finding `zot80.focus`, e.g.
https://github.com/hiddenSymmetries/simsopt/actions/runs/4379941925/jobs/7666434998

akaptano commented on examples/2_Intermediate/permanent_magnet_GPMO.py:
```diff
@@ -0,0 +1,260 @@
+magnet optimization on the MUSE grid.
+
+The script should be run as:
+    mpirun -n 1 python permanent_magnet_GPMO.py
```
That is sufficient, although confusingly, mpirun (actually srun) must be used on NERSC nodes.

akaptano commented on src/simsoptpp/surface.cpp:
```diff
@@ -138,6 +138,44 @@ void Surface<Array>::extend_via_normal(double scale) {
 }
 
+template<class Array>
+void Surface<Array>::extend_via_projected_normal(const vector<double> phi, double scale) {
```
Agreed, it is mainly in the c++ code because it is extremely similar to the extend_via_normal c++ function so I will hold off on changing this to python for now.

akaptano commented on src/simsoptpp/permanent_magnet_optimization.cpp:
```diff
@@ -0,0 +1,1592 @@
+}
+
+// perform a line search
+double cubic_interp(double alpha) {
```
yes, I never finished implementing a fair number of these functions, which were going to be used to do combined stage-1 and stage-2 permanent magnet optimization. I'll delete all these for now, and if someone takes up this project again we can go back to the right commit.

akaptano commented:
@mbkumar Is there a straightforward way I can update the codecov report as I commit new changes? It would be useful to see my progress addressing the unit test lack. Thanks!

mbkumar commented on src/simsopt/field/coil.py:
```diff
@@ -8,7 +8,7 @@
 
 
-__all__ = ['Coil', 'Current', 'coils_via_symmetries',
+__all__ = ['Coil', 'Current', 'ScaledCurrent', 'coils_via_symmetries',
```
ScaledCurrent should not be user exposed. It acts as a proxy class for multiplication. Please revert back this change.

mbkumar commented on src/simsopt/field/magneticfieldclasses.py:
```diff
@@ -502,6 +503,179 @@ def from_dict(cls, d, serial_objs_dict, recon_objs):
+        pointsToVTK(
+            vtkname, ox, oy, oz, data=data
+        )
+
```
With the new JSON framework, you don't have to define `as_dict` and `from_dict` for the `DipoleField` class, and the default methods from `Optimizable `should be sufficient.

mbkumar commented on src/simsopt/field/magneticfieldclasses.py:
```diff
@@ -502,6 +503,179 @@ def from_dict(cls, d, serial_objs_dict, recon_objs):
+        pm_opt: A PermanentMagnetOptimizer object that has already been optimized.
+    """
+
+    def __init__(self, pm_opt=None):
```
You are creating a depening on PermanentMagnetOptimizer class for DipoleField class? Is this really necessary? You are limiting the usage of DipoleField.

akaptano replied:
Agreed, this was unnecessary and went through and fixed this

mbkumar commented on src/simsopt/geo/permanent_magnet_grid.py:
```diff
@@ -0,0 +1,541 @@
+
+        t1 = time.time()
+        # If the inner surface is not specified, make default surface.
+        if (rz_inner_surface is None) and (famus_filename is None):
```
This style (`if (rz_inner_surface is None) and (famus_filename is None)`) used to check for valid arguments is not pythonic. Remove the `famus_filename` argument and create a class method class `from_famus_file` and populate the arguments using famus_file in that classmethod.

mbkumar commented on src/simsopt/geo/permanent_magnet_grid.py:
```diff
@@ -0,0 +1,541 @@
+            d["filename"],
+            d["surface_flag"],
+            d["coordinate_flag"],
+        )
```
as_dict and from_dict need not be implemented.

mbkumar commented:
> @mbkumar Is there a straightforward way I can update the codecov report as I commit new changes? It would be useful to see my progress addressing the unit test lack. Thanks!

@akaptano 
If you are committing changes to this PR, the codecov report can be seen here under "checks".
Could you please explain this PR (motivation, describing the classes, tests, and examples) in our weekly meetings on Friday at 11:00 AM? How about next week Friday?

landreman commented on src/simsopt/geo/permanent_magnet_grid.py:
```diff
@@ -0,0 +1,542 @@
+        ----------
+        famus_filename : TYPE
+            Filename of a FAMUS grid file (a pre-made dipole grid).
+
```
Document `pol_vectors`

landreman reviewed (commented):
We should add the new source files to `docs/source/simsopt.*.rst`. @mbkumar what was the command to do that automatically?

akaptano commented on src/simsopt/util/permanent_magnet_helper_functions.py:
```diff
@@ -0,0 +1,788 @@
+        + CC_WEIGHT * Jccdist \
+        + CS_WEIGHT * Jcsdist \
+        + CURVATURE_WEIGHT * sum(Jcs) \
+        + MSC_WEIGHT * sum(QuadraticPenalty(J, MSC_THRESHOLD) for J in Jmscs)
```
Shouldn't be an issue.

akaptano commented on src/simsopt/solve/permanent_magnet_optimization.py:
```diff
@@ -0,0 +1,396 @@
+            the end result should be independent of
+            the choice of initial condition.
+        algorithm: Convex algorithm to use during relax-and-split.
+            Defaults to the MwPGP algorithm.
```
Only options were MwPGP and SPG, deleted SPG with previous commit so getting rid of this argument

akaptano commented on tests/field/test_magneticfields.py:
```diff
@@ -466,6 +472,218 @@ def test_Dommaschk(self):
+        bs = BiotSavart(coils)
+        bs.set_points(s.gamma().reshape((-1, 3)))
+        Bn = np.sum(bs.B().reshape(nphi, ntheta, 3) * s.unitnormal(), axis=-1)
+        pm_opt = PermanentMagnetGrid(
```
I think this is a holdover from when the class needed a pm_opt object to initialize. Rewriting these without the PM grid now.

landreman commented:
> @mbkumar Is there a straightforward way I can update the codecov report as I commit new changes? It would be useful to see my progress addressing the unit test lack. Thanks!

I believe codecov only updates if all the tests pass. Once the broken tests are fixed, I think we will start to see updates.

landreman commented on src/simsopt/objectives/fluxobjective.py:
```diff
@@ -40,7 +40,10 @@ def __init__(self, surface, field, target=None):
         n = self.surface.normal()
         Bcoil = self.field.B().reshape(n.shape)
-        Btarget = self.target if self.target is not None else []
+        if self.target is not None:
+            Btarget = np.ascontiguousarray(self.target)
+        else:
+            Btarget = []
```
Is there any problem with moving these lines up to the `__init__()` function, replacing `Btarget` -> `self.Btarget`? Florian found that evaluating `J()` was performance-relevant enough to put the next line into C++, so it's possible that having this little bit of code re-run every time `J()` is evaluated might affect performance.

akaptano commented:
@landreman @mbkumar Any idea why the CI is failing? Looks like errors unrelated to my changes.

mbkumar commented:
Merge master into your branch and test on your machine

On Tue, Apr 18, 2023, 8:31 PM Alan Kaptanoglu ***@***.***>
wrote:

> @landreman <https://github.com/landreman> @mbkumar
> <https://github.com/mbkumar> Any idea why the CI is failing? Looks like
> errors unrelated to my changes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hiddenSymmetries/simsopt/pull/296#issuecomment-1514009372>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA62VEBWDBUI47EALJCLYHLXB456VANCNFSM6AAAAAAVTJER6I>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

akaptano commented:
@mbkumar Most recent master is already merged into this branch, and I am unclear how to test this on my machine. Error looks like it was during sudo apt-get install update run by the online CI, which seems totally independent of my device (my Mac doesn't even accept apt-get commands!)

mbkumar commented:
I check the CI failure for your PR tomorrow.

On Tue, Apr 18, 2023, 10:08 PM Alan Kaptanoglu ***@***.***>
wrote:

> @mbkumar <https://github.com/mbkumar> Most recent master is already
> merged into this branch, and I am unclear how to test this on my machine.
> Error looks like it was during sudo apt-get install update run by the
> online CI, which seems totally independent of my device (my Mac doesn't
> even accept apt-get commands!)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hiddenSymmetries/simsopt/pull/296#issuecomment-1514070529>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA62VEDDLKTI37Y2QEKHQ2TXB5JJBANCNFSM6AAAAAAVTJER6I>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

landreman commented:
Sometimes the server for apt-get is randomly down. Manually re-running the
failed jobs usually resolves the problem. I did that just now and all tests
passed.

On Tue, Apr 18, 2023, 9:31 PM Alan Kaptanoglu ***@***.***>
wrote:

> @landreman <https://github.com/landreman> @mbkumar
> <https://github.com/mbkumar> Any idea why the CI is failing? Looks like
> errors unrelated to my changes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hiddenSymmetries/simsopt/pull/296#issuecomment-1514009372>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABH4ZA6VWYEB57HTHOOABVTXB456VANCNFSM6AAAAAAVTJER6I>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

MigMadeira commented on src/simsopt/geo/permanent_magnet_grid.py:
```diff
@@ -0,0 +1,519 @@
+            self.ndipoles = self.dipole_grid_xyz.shape[0]
+            self.pm_phi = np.arctan2(self.dipole_grid_xyz[:, 1], self.dipole_grid_xyz[:, 0])
+            cell_vol = self.dx * self.dy * self.dz 
+            self.m_maxima = B_max * cell_vol / mu0 * np.ones(self.ndipoles)
```
Please consider promoting m_maxima to an input either in `PermanentMagnetGrid` and/or `geo_setup`

akaptano replied:
See above, will fix this shortly.

akaptano commented on src/simsopt/geo/permanent_magnet_grid.py:
```diff
@@ -0,0 +1,519 @@
+        B_max = 1.465  # value used in FAMUS runs for MUSE
+        mu0 = 4 * np.pi * 1e-7
+        cell_vol = M0s * mu0 / B_max
+        self.m_maxima = B_max * cell_vol[self.Ic_inds] / mu0
```
nice suggestion, I'll get this done today!

akaptano commented:
Hi all, finally got this merge to a good place -- unit tests are much better quality + much higher coverage, implemented all of Matt's code suggestions, implemented Miguel's suggestions, so just waiting on any last comments/changes on this.

mbkumar commented:
I'll check it and let you know if I have any comments.

On Thu, Apr 27, 2023, 11:13 AM Alan Kaptanoglu ***@***.***>
wrote:

> Hi all, finally got this merge to a good place -- unit tests are much
> better quality + much higher coverage, implemented all of Matt's code
> suggestions, implemented Miguel's suggestions, so just waiting on any last
> comments/changes on this.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hiddenSymmetries/simsopt/pull/296#issuecomment-1525877738>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA62VEGWKKCALU6ZSYEUZJDXDKEJRANCNFSM6AAAAAAVTJER6I>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mbkumar commented on src/simsopt/geo/permanent_magnet_grid.py:
```diff
@@ -0,0 +1,549 @@
+                'famus_filename variable is set, so a pre-defined grid will be used. '
+                ' so that the following parameters are ignored: '
+                'rz_inner_surface, rz_outer_surface, dr.'
+            )
```
If famus_file has rz_inner_surface, rz_outer_surface, etc, why can't we use them?

mbkumar commented on src/simsopt/geo/permanent_magnet_grid.py:
```diff
@@ -0,0 +1,549 @@
+
+    def _setup_uniform_rz_grid(self):
+        """
+            Initializes a uniform grid in cylindrical coordinates and sets
```
Fix the docstring alignment

mbkumar commented on src/simsopt/objectives/fluxobjective.py:
```diff
@@ -50,8 +52,8 @@ def dJ(self):
         Bcoil_n = np.sum(Bcoil*unitn, axis=2)
-        if self.target is not None:
-            B_n = (Bcoil_n - self.target)
+        if self.Btarget is not None:
```
Line 55 is unnecessary because self.Btarget is set to something always.

mbkumar commented on src/simsopt/solve/permanent_magnet_optimization.py:
```diff
@@ -0,0 +1,422 @@
+
+
+def prox_l0(m, mmax, reg_l0, nu):
+    """
```
Fix the alignment of docstrings.

mbkumar commented on src/simsopt/solve/permanent_magnet_optimization.py:
```diff
@@ -0,0 +1,422 @@
+
+def prox_l1(m, mmax, reg_l1, nu):
+    """
+        Proximal operator for L1 regularization.
```
Fix the alignment of docstrings

mbkumar commented on src/simsopt/solve/permanent_magnet_optimization.py:
```diff
@@ -0,0 +1,422 @@
+
+
+def projection_L2_balls(x, mmax):
+    """
```
Docstring alignment

mbkumar commented on src/simsopt/solve/permanent_magnet_optimization.py:
```diff
@@ -0,0 +1,422 @@
+
+def setup_initial_condition(pm_opt, m0=None):
+    """
+        If an initial guess for the dipole moments is specified,
```
Docstring alignment

mbkumar commented on src/simsopt/util/famus_helpers.py:
```diff
@@ -0,0 +1,548 @@
+"""
```
Docstring alignment

mbkumar commented on src/simsopt/util/famus_helpers.py:
```diff
@@ -0,0 +1,548 @@
+
+        with open(fname, 'r') as f:
+            lines = f.readlines()
+            f.close()
```
Line 25 is unnecessary. When exiting the `with` block, the file is closed automatically.

mbkumar commented on src/simsopt/util/permanent_magnet_helper_functions.py:
```diff
@@ -0,0 +1,655 @@
+
+
+def trace_fieldlines(bfield, label, s, comm, OUT_DIR):
+    """
```
Docstring alignment

mbkumar commented on src/simsopt/util/permanent_magnet_helper_functions.py:
```diff
@@ -0,0 +1,655 @@
+
+def initialize_default_kwargs(algorithm='RS'):
+    """
+        Keywords to the permanent magnet optimizers are now passed
```
docstring alignment

mbkumar commented on src/simsopt/util/polarization_project.py:
```diff
@@ -0,0 +1,545 @@
+
+def faceedge_vectors(theta):
+    """
+        Generates an array of unit vectors for a set of face-edge polarizations
```
Docstring alignment

mbkumar reviewed (changes requested):
I have a few comments. 
1) Can you please introduce type hints and minimize the if-else checks on the arguments?
2) Check the alignment of docstrings. Don't indent them further than the indent of """.
3) This PR and #292 don't go together. We need non-XSIMD implementation for the XSIMD-based C++ code added here. First, let's merge this and then we can update #292.

mbkumar commented:
I'll take one more pass at this PR sometime this weekend.

mbkumar commented:
@akaptano 

Recently, simsopt port without XSIMD dependency is merged into master. A new test  is added to check against this. Make sure your C++ code can work without XSIMD. If required I can take a stab at this.

mbkumar commented:
@akaptano 

Integrated tests (aka examples) are taking more than 3 hours. Can you find a way to reduce the execution time? We limit the iteration numbers and adjust the stopping criteria to speed up the execution in CI. You can leave notes and commented-out code in the examples for users to run at full scale.

akaptano commented on src/simsoptpp/dipole_field.h:
```diff
@@ -0,0 +1,22 @@
+
+Array dipole_field_dA(Array& points, Array& m_points, Array& m);
+
+std::tuple<Array, Array> dipole_field_Bn(Array& points, Array& m_points, Array& unitnormal, int nfp, int stellsym, Array& b, std::string coordinate_flag="cartesian", double R0=0.0);
+
+std::tuple<Array, Array> define_a_uniform_cylindrical_grid_between_two_toroidal_surfaces(Array& phi, Array& normal_inner, Array& normal_outer, Array& dipole_grid_rz, Array& r_inner, Array& r_outer, Array& z_inner, Array& z_outer);
+
+Array define_a_uniform_cartesian_grid_between_two_toroidal_surfaces(Array& normal_inner, Array& normal_outer, Array& xyz_uniform, Array& xyz_inner, Array& xyz_outer);
```
Yes, done now

mbkumar commented on src/simsopt/solve/permanent_magnet_optimization.py:
```diff
@@ -0,0 +1,415 @@
+            ArbVec_backtracking (same as above but w/ backtracking).
+            Easiest to use is baseline but most effective algorithm
+            is ArbVec_backtracking.
+        algorithm_kwargs: dict
```
Please use `kwargs`, which is the canonical python argument, in place of `algorithm_kwargs`.

mbkumar commented on src/simsopt/solve/permanent_magnet_optimization.py:
```diff
@@ -0,0 +1,415 @@
+    if not hasattr(pm_opt, "A_obj"):
+        raise ValueError(
+            "The PermanentMagnetClass needs to use geo_setup() or "
+            "geo_setup_from_famus() before calling optimization routines."
```
Another reason to make `geo_setup_from_famus` and `geo_setup` class methods.

mbkumar commented on src/simsopt/solve/permanent_magnet_optimization.py:
```diff
@@ -0,0 +1,415 @@
+        reg_l2 = algorithm_kwargs["reg_l2"]
+        algorithm_kwargs.pop("reg_l2")
+    else:
+        reg_l2 = 0.0
```
Lines 324 to 328 can be replaced with

`reg_l2 = algorithm_kwargs.pop("reg_l2", 0.0)`

mbkumar commented on src/simsopt/solve/permanent_magnet_optimization.py:
```diff
@@ -0,0 +1,415 @@
+        reg_l2 = 0.0
+
+    # check that algorithm can generate K binary dipoles
+    if "K" in algorithm_kwargs.keys():
```
This is old school.
 
`if "K" in algorithm_kwargs:` should be sufficient. Noticed similar usage elsewhere.

mbkumar commented on src/simsopt/solve/permanent_magnet_optimization.py:
```diff
@@ -0,0 +1,415 @@
+                'Initial dipole guess is incorrect shape --'
+                ' guess must be 1D with shape (ndipoles * 3).'
+            )
+        m0_temp = projection_L2_balls(
```
One typically breaks arguments into subsequent lines when there are too many arguments, or the line goes past 80 chars. It is distracting for small functions.

mbkumar commented on src/simsopt/util/famus_helpers.py:
```diff
@@ -0,0 +1,571 @@
+        filename: a FOCUS file
+    """
+
+    propNames = ['type', 'symm', 'coilname', 'ox', 'oy', 'oz', 'Ic', 'M_0', \
```
Linebreak is redundant.

mbkumar commented on src/simsopt/util/famus_helpers.py:
```diff
@@ -0,0 +1,571 @@
+
+    def read_from_file(self, filename, downsample):
+
+        focusfile = open(filename, 'r')
```
Please use `with open...` context manager.

mbkumar commented on src/simsopt/util/permanent_magnet_helper_functions.py:
```diff
@@ -0,0 +1,716 @@
+    return coils, base_currents, ncoils
+
+
+def coil_optimization(s, bs, base_curves, curves, OUT_DIR):
```
Please use lowercase for OUT_DIR in all places. Also, think about supplying a default argument such as None to use the current directory.

mbkumar commented on src/simsopt/util/permanent_magnet_helper_functions.py:
```diff
@@ -0,0 +1,716 @@
+    ################################################################################
+    """)
+    res = minimize(fun, dofs, jac=True, method='L-BFGS-B', options={'maxiter': MAXITER, 'maxcor': 300}, tol=1e-15)
+    curves_to_vtk(curves, OUT_DIR + f"curves_opt")
```
Make it portable. `Path(out_dir) / "curves_opt"`

mbkumar commented on src/simsopt/util/permanent_magnet_helper_functions.py:
```diff
@@ -0,0 +1,716 @@
+    return coils, base_currents, ncoils
+
+
+def coil_optimization(s, bs, base_curves, curves, OUT_DIR):
```
Is `simsopt.solve.permanent_magnet_optimization` a better place for this function?

akaptano replied:
I don't think so. This for optimizing any existing coils, which is done before optimizing any permanent magnets (and the optimization problems are quite different).

mbkumar commented on src/simsopt/util/polarization_project.py:
```diff
@@ -0,0 +1,548 @@
+to the PM4Stell team and Ken Hammond for his consent to use this file
+and work with the permanent magnet branch of SIMSOPT.
+"""
+__all__ = ['faceedge_vectors', 'facecorner_vectors', 'edge_triplet', 
```
Do you want to expose all the functions?

mbkumar commented on tests/geo/test_pm_grid.py:
```diff
@@ -0,0 +1,704 @@
+        )
+
+        with self.assertRaises(ValueError):
+            PermanentMagnetGrid(
```
Use single line

mbkumar reviewed (changes requested):
Hi Alan,

Please go throw my comments and make the required changes. The changes needed will help in long-term maintainability.

akaptano commented on src/simsopt/solve/permanent_magnet_optimization.py:
```diff
@@ -0,0 +1,415 @@
+import simsoptpp as sopp
+import warnings
+
+__all__ = ['relax_and_split', 'GPMO', 'prox_l0', 'prox_l1', 'setup_initial_condition']
```
permanent_magnet_QA.py uses the relax_and_split algorithm. But the other ones I agree, and this has been fixed.

akaptano commented:
Pushing up your request changes shortly, thanks!

mbkumar reviewed (changes requested):
Hello Alan,

Thanks for making the required changes. I am almost ready to approve this PR. I have a few requests in a couple of places. Could you please address them?

Thank you,
Bharat

akaptano commented:
Pushing up these changes in a moment, thanks!

mbkumar reviewed (approved):
Hello Alan,

Thanks for the changes. I am signing off on this PR.

mbkumar reviewed (approved):
Hello Alan,

Thanks for the changes. I am signing off on this PR.
